src/Type/MenuItem/MenuItemType.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+namespace WPGraphQL\Type\MenuItem;
+
+use GraphQLRelay\Relay;
+use WPGraphQL\Type\WPObjectType;
+use WPGraphQL\Types;
+use WPGraphQL\Type\MenuItem\Connection\MenuItemConnectionDefinition;
+
+/**
+ * Class MenuItemType
+ *
+ * @package WPGraphQL\Type\MenuItem
+ * @since 0.0.29
+ */
+class MenuItemType extends WPObjectType {
+
+	/**
+	 * Holds the type name
+	 *
+	 * @var string $type_name
+	 */
+	private static $type_name;
+
+	/**
+	 * This holds the field definitions
+	 *
+	 * @var array $fields
+	 */
+	private static $fields;
+
+	/**
+	 * MenuItemType constructor.
+	 */
+	public function __construct() {
+
+		self::$type_name = 'MenuItem';
+
+		$config = [
+			'name'        => self::$type_name,
+			'description' => __( 'Navigation menu items are the individual items assigned to a menu. These are rendered as the links in a navigation menu.', 'wp-graphql' ),
+			'fields'      => self::fields(),
+		];
+
+		parent::__construct( $config );
+
+	}
+
+	/**
+	 * This defines the fields that make up the MenuItemType
+	 *
+	 * @return array|\Closure|null
+	 */
+	private static function fields() {
+
+		if ( null === self::$fields ) {
+
+			$local_id_name = lcfirst( self::$type_name ) . 'Id';
+
+			self::$fields = function() {
+				$fields = [
+					'id' => [
+						'type'        => Types::non_null( Types::id() ),
+						'description' => __( 'Relay ID of the menu item.', 'wp-graphql' ),
+						'resolve'     => function( \WP_Post $menu_item ) {
+							return ! empty( $menu_item->ID ) ? Relay::toGlobalId( self::$type_name, $menu_item->ID ) : null;
+						},
+					],
+					'childItems' => MenuItemConnectionDefinition::connection(),
+					'connectedObject' => [
+						'type'        => Types::menu_item_object_union(),
+						'description' => __( 'The object connected to this menu item.', 'wp-graphql' ),
+						'resolve' => function( \WP_Post $menu_item ) {
+							$object_id   = get_post_meta( $menu_item->ID, '_menu_item_object_id', true );
+							$object_type = get_post_meta( $menu_item->ID, '_menu_item_type', true );
+
+							// By default, resolve to the menu item itself. This is the
+							// case for custom links.
+							$resolved_object = $menu_item;
+
+							switch ( $object_type ) {
+								// Post object
+								case 'post_type':
+									$resolved_object = get_post( $object_id );
+									break;
+
+								// Taxonomy term
+								case 'taxonomy':
+									$resolved_object = get_term( $object_id );
+									break;
+							}
+
+							/**
+							 * Allow users to override how nav menu items are resolved.
+							 * This is useful since we often add taxonomy terms to menus
+							 * but would prefer to represent the menu item in other ways,
+							 * e.g., a linked post object (or vice-versa).
+							 */
```

Comment:
This resolver should probably include the other args that come along (`$args`, `$context`, `$info`) then pass those to this filter so anyone hooking to this filter has that extra resolve info to use if needed. 

Might be good to pass the $object_id and $object_type to the filter as well